external/types/core.d.ts

```diff
@@ -0,0 +1,227 @@
+    [file: string]: FileSpec;
+  }
+
+  type DestinationTarget = any[];
+
+  type Destination = {
+    [destination: string]: any[];
+  };
+
+  //
+  // src/core/primitives.js
+  //
+  interface Ref {
+    num: number;
+    gen: number;
+  }
+
+  // Interning names
+  interface Name {
+    name: string;
+  }
+
+  interface StringMap {
+    [key: string]: string;
+  }
+
+  //
+  // src/core/annotation.js
+  //
+  interface Annotation {
+    annotationFlags: AnnotationFlag;
+    annotationType: AnnotationType;
+    borderStyle: AnnotationBorderStyle;
+    color: Color | null;
+    hasAppearance: boolean;
+    id: string;
+    rect: Rect;
+    subtype: AnnotationSubtype | null;
+
+    // For popup annotations
+    hasPopup?: boolean;
+    title?: string;
+    contents?: string;
+  }
+
+  /**
+   * Contains all data regarding an annotation's border style.
+   */
+  class AnnotationBorderStyle {
+    width: number;
+    style: AnnotationBorderStyleType;
+    dashArray: number[];
+    horizontalCornerRadius: number;
+    verticalCornerRadius: number;
+  }
+
+  interface WidgetAnnotation extends Annotation {
+    annotationType: AnnotationType.WIDGET;
+    fieldName: string;
+    fieldType: string;
+    fieldValue: string | string[] | null;
+    alternativeText: string;
+    defaultAppearance: string;
+    fieldFlags: number;
+    readOnly: boolean;
+  }
+  interface TextWidgetAnnotation extends WidgetAnnotation {
+    fieldValue: string;
+    textAlignment: number | null;
+    maxLen: number | null;
+    multiLine: boolean;
+    comb: boolean;
+  }
+  interface ButtonWidgetAnnotation extends WidgetAnnotation {
+    checkBox: boolean;
+    radioButton: boolean;
+    pushButton: boolean;
+    fieldValue: string | null;
+    buttonValue?: string | null;
+  }
+  interface ChoiceWidgetAnnotation extends WidgetAnnotation {
+    options: ChoiceOption[];
+    fieldValue: string[];
+    combo: boolean;
+    multiSelect: boolean;
+  }
+  interface ChoiceOption {
+      exportValue: string;
+      displayValue: string;
+  }
+  interface TextAnnotation extends Annotation {
+    annotationType: AnnotationType.TEXT;
+    name: string;
+  }
+  interface LinkAnnotation extends Annotation {
+    annotationType: AnnotationType.LINK;
+  }
+  interface PopupAnnotation extends Annotation {
+    annotationType: AnnotationType.POPUP;
+    parentType: string | null;
+    parentId: string;
+    title: string;
+    contents: string;
+    color: Color | null;
+  }
+  interface LineAnnotation extends Annotation {
+    annotationType: AnnotationType.LINE;
+    lineCoordinates: Rect;
+  }
+  interface SquareAnnotation extends Annotation {
+    annotationType: AnnotationType.SQUARE;
+  }
+  interface CircleAnnotation extends Annotation {
+    annotationType: AnnotationType.CIRCLE;
+  }
+  interface PolylineAnnotation extends Annotation {
+    annotationType: AnnotationType.POLYLINE;
+    vertices: Vertex[];
+  }
+  interface PolygonAnnotation extends Annotation {
```

Comment:
The annotation layer changes are in https://github.com/mozilla/pdf.js/blob/master/src/core/annotation.js (core) and https://github.com/mozilla/pdf.js/blob/master/src/display/annotation_layer.js (display). I think `InkAnnotation` is missing, but maybe there is more.

Reply:
Done, see b74c5fd.